test/hotspot/jtreg/runtime/os/TestTracePageSizes.java

```diff
@@ -0,0 +1,279 @@
+    // transparent huge pages. The pattern is not greedy and will
+    // match as little as possible so each "segment" in the file
+    // will generate a match.
+    private static void parseSmaps() throws Exception {
+        Pattern smapsPattern = Pattern.compile(RangeWithPageSize.smapsPatternString, Pattern.DOTALL);
+        Scanner smapsScanner = new Scanner(new File("/proc/self/smaps"));
+        // Find all memory segments in the smaps-file.
+        smapsScanner.findAll(smapsPattern).forEach(mr -> {
+            String start = mr.group(1);
+            String end = mr.group(2);
+            String ps = mr.group(3);
+            String vmFlags = mr.group(4);
+
+            // Create a range given the match and add it to the list.
+            RangeWithPageSize range = new RangeWithPageSize(start, end, ps, vmFlags);
+            ranges.add(range);
+            debug("Added range: " + range);
+        });
+        smapsScanner.close();
+    }
+
+    // Search for a range including the given address.
+    private static RangeWithPageSize getRange(String addr) {
+        long laddr = Long.decode(addr);
+        for (RangeWithPageSize range : ranges) {
+            if (range.includes(laddr)) {
+                return range;
+            }
+        }
+        return null;
+    }
+
+    // Helper to get the page size in KB given a page size parsed
+    // from log_info(pagesize) output.
+    private static long pageSizeInKB(String pageSize) {
+        String value = pageSize.substring(0, pageSize.length()-1);
+        String unit = pageSize.substring(pageSize.length()-1);
+        long ret = Long.parseLong(value);
+        if (unit.equals("K")) {
+            return ret;
+        } else if (unit.equals("M")) {
+            return ret * 1024;
+        } else if (unit.equals("G")) {
+            return ret * 1024 * 1024;
+        }
+        return 0;
+    }
+
+    // The test needs to be run with:
+    //  * -Xlog:pagesize:ps-%p.log - To generate the log file parsed
+    //    by the test itself.
+    //  * -XX:+AlwaysPreTouch - To make sure mapped memory is touched
+    //    so the relevant information is recorded in the smaps-file.
+    public static void main(String args[]) throws Exception {
+        // Check if debug printing is enabled.
+        if (args.length > 0 && args[0].equals("-debug")) {
+            debug = true;
+        } else {
+            debug = false;
+        }
+
+        // Parse /proc/self/smaps to compare with values logged in the VM.
+        parseSmaps();
+
+        // Setup patters for the JVM page size logging.
+        String traceLinePatternString = ".*base=(0x[0-9A-Fa-f]*).*page_size=([^ ]+).*";
+        Pattern traceLinePattern = Pattern.compile(traceLinePatternString);
+
+        // The test needs to be run with page size logging printed to ps-$pid.log.
+        Scanner fileScanner = new Scanner(new File("./ps-" + ProcessHandle.current().pid() + ".log"));
```

Comment:
So we scan the log file while it is being written by ourselves? This seems slightly unsafe. Are we sure it has been completely flushed at this point? I'm mainly concerned about things like running this test in a current dir on NFS or somesuch.

An alternative would be to let the testee write the log file and print out a copy of /proc/self/smaps to stdout, then to scan and compare in a separate process. 

But if you think this is safe enough, I'm fine with this too.

Reply:
Not really while it is being written, since all the output from `-Xlog:pagesize` happens before the java test is executed and from what I can tell the log-file should be properly flushed.

As you say doing the scan and compare in a separate process afterwards would make any concurrent modification issue and flush issue go away. But it would also slightly complicate the test and make it depend on the test library. For debugging it is very nice to be able to run the test standalone and that is a bit harder when depending on the test lib.

Until proven unsafe I would prefer to leave it as is.

Reply:
I'd say so too. If that pattern works, I may use it in other places too.